locale/en/docs/guides/dont-block-the-event-loop.md

```diff
@@ -0,0 +1,435 @@
+Node scales well, sometimes better than more heavyweight approaches like Apache.
+The secret to Node's scalability is that it uses a small number of threads to handle many clients.
+If Node can make do with fewer threads, then it can spend more of your system's time and memory working on clients rather than on paying overheads for threads.
+
+Here's a good rule of thumb for keeping your Node server speedy:
+*Node is only fast when the work associated with each client at any given time is "small"*.
+
+This applies to work both on the Event Loop and on the Worker Pool.
+
+## Why should I avoid blocking the Event Loop and the Worker Pool?
+Node uses a small number of threads to handle many clients.
+While a thread is working on behalf of one client, it can't work for any other clients.
+
+This provides two motivations for blocking neither the Event Loop nor the Worker Pool.
+Since both the Event Loop and the Workers respond to events, we will refer to these threads generally as *Event Handlers* or *Event-Handler Threads*.
+All threads in Node are Event Handlers: one Event Loop and several Workers.
+
+1. Performance: If you regularly perform heavyweight tasks on an Event Handler, the *throughput* (requests/second) of your server will suffer.
+2. Security: If it is possible that for certain input you might block an Event Handler, a malicious client might be able to figure out the "evil input", make your code block, and keep you from working on other clients. This would be a Denial of Service attack.
+
+## A quick review of Node's architecture
+
+Node uses the Event-Driven Architecture: it has an Event Loop for orchestration and a Worker Pool for expensive tasks.
+
+## Queues and Cooperative Multitasking
+Both the Event Loop and the Worker Pool use *queues*, which are data structures with one or more producers who place items into the queue, and one or more consumers who remove items from the queue.
+
+The Event Loop's queue is full of pending events.
+When the Event Loop sees a new event, it invokes its callback, which is JavaScript code that sees your Node application.
+The Event Loop is the only consumer of this queue.
+The Node runtime produces events for this queue.
+Specifically, Node asks libuv to monitor file descriptors using [epoll](http://man7.org/linux/man-pages/man7/epoll.7.html)), and the Event Loop's queue consists of the ready descriptors.
+
+The Worker Pool's queue is full of pending Tasks.
+When a Worker sees a new Task, it invokes its callback, which is C++ code that sits on the C++ side of the Node core modules.
+The consumers of this queue are the Workers in the Worker Pool, who all compete to work on the next Task.
+Most of the time, the Event Loop is the only producer of Tasks for the Worker Pool.
+
+It is crucial to understand that this queue-based approach means that the Event Loop and the Workers rely on [cooperative multitasking](https://en.wikipedia.org/wiki/Cooperative_multitasking).
```

Comment:
this isn't the common understanding of cooperative multi-tasking. coop tasking means that the running thread can suspend itself, but node doesn't do this, ever (on main loop or in pool). it runs to completion, which is not cooperative multitasking

Reply:
I couldn't think of a better way to describe this other than saying "queue-based". Should I stick with that? Is there a better term?

Counterpoint, though -- queue-based is "cooperative multitasking" in the sense that a callback/task yields precisely when it completes.

Reply:
I think that other than the queue to the uv work pool (which is actually a queue), I don't think queues should be mentioned. Not sure what a good replacement is, sorry! I can see the ghostly parallels, both node and co-op multitasking systems require careful consideration of how much work you do, and how you construct your app, but the analogy breaks down fast, because cooperative multi-tasking has a fairly agreed upon definition, and its not what node does, so better stay away from it, IMO.